primitives/src/transaction.rs

```diff
@@ -349,6 +351,388 @@ impl Encodable for Transaction {
     }
 }
 
+/// The decoder for the [`Transaction`] type.
+#[cfg(feature = "alloc")]
+pub struct TransactionDecoder {
+    state: TransactionDecoderState,
+}
+
+#[cfg(feature = "alloc")]
+impl TransactionDecoder {
+    /// Constructs a new [`TransactionDecoder`].
+    pub fn new() -> Self { Self { state: TransactionDecoderState::Version(VersionDecoder::new()) } }
+}
+
+#[cfg(feature = "alloc")]
+impl Default for TransactionDecoder {
+    fn default() -> Self { Self::new() }
+}
+
+/// The state of the transiting decoder.
+#[cfg(feature = "alloc")]
+pub enum TransactionDecoderState {
```

Comment:
In 15b4f4c43bc790e2e86797f5dee56f7af588ba54, do we have to expose `TransactionDecoderState`/`Attempt`/`IsSegwit`? Would be nice to give us the flexibility to tweak them in the future if we can hide them.

Reply:
Oh BOOM! I thought I tried that before, that's nice!